File: test/local/check.c
#include "mpi.h"
#include "fti.h"
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <unistd.h>

#define N 100000
#define CNTRLD_EXIT 10
#define RECOVERY_FAILED 20
#define DATA_CORRUPT 30
#define BACKUP_DATA_CORRUPT 40
#define FTI_ERROR 50
#define MPI_ERROR 60
#define KEEP 2
#define RESTART 1
#define INIT 0

/** 
 * function prototypes
 */

void init_arrays(double* A, double* B, size_t asize);

void vecmulc(double* A, double* B, size_t asize);

int validify(double* A, double* B, size_t asize);

int write_data(double* B, size_t* asize, int rank);

int read_data(double* B_chk, size_t* asize_chk, int rank, size_t asize);

/**
 * main
 */ 

int main(int argc, char* argv[]) {

    unsigned char parity, crash, level, state, diff_sizes;
    int FTI_APP_RANK, result, tmp, success = 1;
    double *A, *B, *B_chk;

    size_t asize, asize_chk;

    srand(time(NULL));

    MPI_Init(&argc, &argv);
    FTI_Init(argv[1], MPI_COMM_WORLD);
    
    crash = atoi(argv[2]);
    level = atoi(argv[3]);
    diff_sizes = atoi(argv[4]);

    MPI_Comm_rank(FTI_COMM_WORLD,&FTI_APP_RANK);
    
    asize = N;
    
    if (diff_sizes) {
        parity = FTI_APP_RANK%7;

        switch (parity) {

            case 0:
                asize = N;
                break;

            case 1:
                asize = 2*N;
                break;

            case 2:
                asize = 3*N;
                break;

            case 3:
                asize = 4*N;
                break;

            case 4:
                asize = 5*N;
                break;

            case 5:
                asize = 6*N;
                break;

            case 6:
                asize = 7*N;
                break;

        }
    }

    A = (double*) malloc(asize*sizeof(double));
    B = (double*) malloc(asize*sizeof(double));

    FTI_Protect(0, A, asize, FTI_DBLE);
    FTI_Protect(1, B, asize, FTI_DBLE);
    FTI_Protect(2, &asize, 1, FTI_INTG);
    
    state = FTI_Status();

    if (state == INIT) {
        init_arrays(A, B, asize);
        write_data(B, &asize, FTI_APP_RANK);
        MPI_Barrier(FTI_COMM_WORLD);
        FTI_Checkpoint(1,level);
        sleep(2);
        if (crash && FTI_APP_RANK == 0) { 
            exit(CNTRLD_EXIT);
        }
    }
    
    if ( state == RESTART || state == KEEP ) {
        result = FTI_Recover();
        if (result != FTI_SCES) {
            exit(RECOVERY_FAILED);
        }
        B_chk = (double*) malloc(asize*sizeof(double));
        result = read_data(B_chk, &asize_chk, FTI_APP_RANK, asize);
        MPI_Barrier(FTI_COMM_WORLD);
        if (result != 0) {
            exit(DATA_CORRUPT);
        }
    }
    
    /*
     * on INIT, B is initialized randomly
     * on RESTART or KEEP, B is recovered and must be equal to B_chk
     */

    vecmulc(A, B, asize);

    if (state == RESTART || state == KEEP) {
        result = validify(A, B_chk, asize);
        result += (asize_chk == asize) ? 0 : -1;
        MPI_Allreduce(&result, &tmp, 1, MPI_INT, MPI_SUM, FTI_COMM_WORLD);
        result = tmp;
        free(B_chk);
    }
    
    free(A);
    free(B);
    
    if (FTI_APP_RANK == 0 && (state == RESTART || state == KEEP)) {
        if (result == 0) {
            printf("[SUCCESSFULL]\n");
        } else {
            printf("[NOT SUCCESSFULL]\n");
            success=0;
        }
    }
    
    MPI_Barrier(FTI_COMM_WORLD);
    FTI_Finalize();
    MPI_Finalize();

    if (success == 1) 
        return 0;
    else 
        exit(DATA_CORRUPT);

}

/**
 * function definitions
 */

void init_arrays(double* A, double* B, size_t asize) {
    int i;
    double r;
    for (i = 0; i< asize; i++) {
        A[i] = 1.0;
        B[i] = ((double)rand()/RAND_MAX)*5.0; 
    }
}

void vecmulc(double* A, double* B, size_t asize) {
    int i;
    for (i=0; i<asize; i++) {
        A[i] = A[i]*B[i];
    }
}

int validify(double* A, double* B, size_t asize) {
    int i;
    for (i=0; i<asize; i++) {
        if (A[i] != B[i]){
            return -1;
        }
    }
    return 0;
}

int write_data(double* B, size_t *asize, int rank) {
    char str[256];
    sprintf(str, "chk/check-%i.tst", rank);
    FILE* f = fopen(str, "wb");
    size_t written = 0;

    fwrite( (void*) asize, sizeof(size_t), 1, f);

    while ( written < (*asize) ) {
        written += fwrite( (void*) B, sizeof(double), (*asize), f);
    }

    fclose(f);

    return 0;
}

int read_data(double* B_chk, size_t *asize_chk, int rank, size_t asize) {
    char str[256];
    sprintf(str, "chk/check-%i.tst", rank);
    FILE* f = fopen(str, "rb");
    size_t read = 0;

    fread( (void*) asize_chk, sizeof(size_t), 1, f);
    if ((*asize_chk) != asize) {
        printf("[ERROR -%i] : wrong dimension 'asize' -- asize: %zd, asize_chk: %zd\n", rank, asize, *asize_chk);
        fflush(stdout);
        return -1;
    }
    while ( read < (*asize_chk) ) {
        read += fread( (void*) B_chk, sizeof(double), (*asize_chk), f);
    }

    fclose(f);

    return 0;
}

File: test/local/check.f90
!> @author Kai Keller (kellekai@gmx.de)
!> @date   June, 2017
!> @brief  Unitary test program for FTI Fortran
!! 
!! The purpose of this code is to test the checkpoint
!! and restart mechanism in FTI for any case.
!!
!! [ USAGE ]
!! 
!! ./PROGRAM ARG1=config ARG2=crash ARG3=level ARG4=diff
!!
!> @param ARG1 relative/absolute path to FTI config file 
!> @param ARG2 1/0 (on/off) - simulate failure  
!> @param ARG3 checkpoint level (1 - 4) 
!> @param ARG4 1/0 (on/off) - enable/disable different ckpt sizes  

MODULE CHECK_FUNCTIONS
   IMPLICIT NONE
   PRIVATE

   PUBLIC INIT_ARRAYS, VALIDIFY, WRITE_DATA, READ_DATA, &
      BUFFER, IERROR

   INTEGER, PARAMETER      :: BUFFER = 256
   INTEGER                 :: IERROR

   CONTAINS
   
   SUBROUTINE INIT_ARRAYS ( A, B )
      REAL, INTENT(OUT)    :: A(:), B(:)
      INTEGER              :: VALUES(1:8), K, I=1
      INTEGER, DIMENSION(:), ALLOCATABLE :: SEED
   
      CALL DATE_AND_TIME(VALUES=VALUES)
      CALL RANDOM_SEED(SIZE=K)
      ALLOCATE(SEED(1:K))
      SEED(:) = VALUES(8)
      CALL RANDOM_SEED(PUT=SEED)
   
      A = 1.0
   
      DO WHILE (I <= SIZE(B))
      CALL RANDOM_NUMBER(B(I))
      I = I+1
      END DO
   END SUBROUTINE 
   
   SUBROUTINE VALIDIFY ( A, B_CHK, ASIZE, RES )
      REAL, INTENT(IN)           :: A(:), B_CHK(:)
      INTEGER (8), INTENT(IN)    :: ASIZE
      INTEGER, INTENT(OUT)       :: RES
      INTEGER                    :: I = 1
   
      DO
      IF ( A(I) /= B_CHK(I) ) THEN
         RES = -1
         EXIT
      END IF
      IF ( I == ASIZE ) EXIT
      I = I+1
      END DO
   END SUBROUTINE
   
   SUBROUTINE WRITE_DATA (B, ASIZE, RANK)
      INTEGER, INTENT(IN)           :: RANK
      INTEGER (8), INTENT(IN)       :: ASIZE
      REAL, INTENT(IN)              :: B(:)
      CHARACTER (LEN=BUFFER)        :: STR
      INTEGER                       :: RECLEN1, RECLEN2, RECLEN, I
   
      WRITE ( STR, '(a,I5.5,a)') 'chk/check-', RANK, ".tst"
   
      INQUIRE ( IOLENGTH=RECLEN1 ) ( B(I), I=1,ASIZE )
      INQUIRE ( IOLENGTH=RECLEN2 ) ASIZE
      RECLEN = RECLEN1 + RECLEN2
   
      OPEN ( 55, FILE=STR, STATUS='REPLACE', IOSTAT=IERROR, FORM='UNFORMATTED', &
         ACCESS='DIRECT', RECL=RECLEN, ACTION='WRITE' )
   
      WRITE ( 55, REC = 1 ) ASIZE
      WRITE ( 55, REC = 2 ) B
   
      CLOSE ( 55 )
   END SUBROUTINE
   
   SUBROUTINE READ_DATA ( B_CHK, ASIZE_CHK, RANK, ASIZE, RES )
      INTEGER, INTENT(IN)         :: RANK
      INTEGER (8), INTENT(IN)     :: ASIZE
      INTEGER (8)                 :: ASIZE_CHK
      REAL, INTENT(OUT)           :: B_CHK(:)
      CHARACTER (LEN=BUFFER)      :: STR
      INTEGER                     :: RECLEN1, RECLEN2, RECLEN, I
      INTEGER, INTENT(OUT)        :: RES
   
      WRITE ( STR, '(a,I5.5,a)') 'chk/check-', RANK, ".tst"
   
      INQUIRE ( IOLENGTH=RECLEN1 ) ASIZE_CHK
      INQUIRE ( IOLENGTH=RECLEN2 ) ( B_CHK(I), I=1,ASIZE )
      RECLEN = RECLEN1+RECLEN2
   
      OPEN ( 55, FILE=STR, STATUS='UNKNOWN', IOSTAT=IERROR, FORM='UNFORMATTED', &
         ACCESS='DIRECT', RECL=RECLEN, ACTION='READ' )
   
      READ ( 55, REC = 1 ) ASIZE_CHK
   
      IF ( ASIZE_CHK /= ASIZE ) THEN
         WRITE(*,'(A, I5.5, A)') "[ERROR -", RANK, "] : wrong dimension"
         RES = -1
         CLOSE ( 55 )
      ELSE 
         READ ( 55, REC = 2 ) B_CHK
         CLOSE ( 55 )
      END IF
   END SUBROUTINE

END MODULE

PROGRAM CHECK
   USE MPI
   USE FTI
   USE CHECK_FUNCTIONS

   IMPLICIT NONE

   !**** CONSTANTS
   INTEGER, PARAMETER      :: N = 10
   INTEGER, PARAMETER      :: CNTRLD_EXIT = 10
   INTEGER, PARAMETER      :: RECOVERY_FAILED = 20
   INTEGER, PARAMETER      :: DATA_CORRUPT = 30
   INTEGER, PARAMETER      :: KEEP = 2
   INTEGER, PARAMETER      :: RESTART = 1
   INTEGER, PARAMETER      :: INIT = 0

   !**** VARIABLE DEFINITIONS
   INTEGER                 :: CRASH, LEVEL, DIFF_SIZES
   INTEGER                 :: RES, TMP, PAR, STATE, SCES = 1
   INTEGER                 :: FTI_APP_RANK, INVALID
   INTEGER (8)             :: ASIZE_CHK
   INTEGER, TARGET         :: FTI_COMM_WORLD
   CHARACTER (BUFFER)      :: CF
   CHARACTER (1)           :: CRASH_CHAR, LEVEL_CHAR, DIFF_SIZES_CHAR

   !**** DYNAMIC ARRAYS
   REAL, DIMENSION(:), ALLOCATABLE            :: B_CHK


   !**** FTI PROTECTED VARIABLES
   REAL, TARGET, DIMENSION(:), ALLOCATABLE    :: A, B 
   INTEGER (8), TARGET     :: ASIZE

   !**** POINTER FOR FTI PROTECTED VARS
   REAL, POINTER           :: A_PTR(:), B_PTR(:) 
   INTEGER (8), POINTER    :: ASIZE_PTR

   !**** GET ARGUMENTS FROM COMMAND LINE 
   CALL GETARG(1, CF)
   CALL GETARG(2, CRASH_CHAR)
   CALL GETARG(3, LEVEL_CHAR)
   CALL GETARG(4, DIFF_SIZES_CHAR)
   READ (CRASH_CHAR(1:1),'(I4)') CRASH
   READ (LEVEL_CHAR(1:1),'(I4)') LEVEL
   READ (DIFF_SIZES_CHAR(1:1),'(I4)') DIFF_SIZES

   !**** MAIN PROGRAM STARTS HERE
   CALL MPI_INIT(IERROR)
   FTI_COMM_WORLD = MPI_COMM_WORLD
   CALL FTI_INIT(CF, FTI_COMM_WORLD, IERROR)

   CALL MPI_COMM_RANK(FTI_COMM_WORLD, FTI_APP_RANK, IERROR)

   ASIZE = N 

   IF ( DIFF_SIZES == 1 ) THEN

      PAR = MOD( FTI_APP_RANK, 7 )
      
      SELECT CASE (PAR) 
      
         CASE ( 0 )
            ASIZE = N;
         CASE ( 1 )
            ASIZE = 2*N;
         CASE ( 2 )
            ASIZE = 3*N;
         CASE ( 3 )
            ASIZE = 4*N;
         CASE ( 4 )
            ASIZE = 5*N;
         CASE ( 5 )
            ASIZE = 6*N;
         CASE ( 6 )
            ASIZE = 7*N;
      
      END SELECT

   END IF

   ALLOCATE( A(ASIZE), STAT=IERROR )
   ALLOCATE( B(ASIZE), STAT=IERROR )

   A_PTR => A
   B_PTR => B
   ASIZE_PTR => ASIZE

   CALL FTI_STATUS( STATE );

   CALL FTI_PROTECT(0, A_PTR, IERROR);
   CALL FTI_PROTECT(1, B_PTR, IERROR);
   CALL FTI_PROTECT(2, ASIZE_PTR, IERROR);

   IF (STATE == INIT) THEN
      CALL INIT_ARRAYS ( A, B )
      CALL WRITE_DATA (B, ASIZE, FTI_APP_RANK);
      CALL MPI_BARRIER ( FTI_COMM_WORLD, IERROR )
      CALL FTI_CHECKPOINT ( 1, LEVEL, IERROR )
      CALL SLEEP(2)
      IF ( CRASH == 1 .AND. FTI_APP_RANK == 0 ) THEN 
         CALL EXIT ( CNTRLD_EXIT )
      END IF
   END IF

   IF ( STATE == RESTART .OR. STATE == KEEP ) THEN
      CALL FTI_RECOVER ( IERROR )
      IF (IERROR /= 0) THEN
         CALL EXIT ( RECOVERY_FAILED )
      END IF
      ALLOCATE ( B_CHK(ASIZE), STAT=IERROR )
      CALL READ_DATA ( B_CHK, ASIZE_CHK, FTI_APP_RANK, ASIZE, IERROR )
      CALL MPI_BARRIER ( FTI_COMM_WORLD, IERROR )
      IF ( IERROR /= 0 ) THEN
         CALL EXIT ( DATA_CORRUPT )
      END IF
   END IF

   !**** CALL VECMULT

   A = A*B

   IF (STATE == RESTART .OR. STATE == KEEP) THEN
      CALL VALIDIFY ( A, B_CHK, ASIZE, INVALID )
      CALL MPI_Allreduce( INVALID, TMP, 1, MPI_INT, MPI_SUM, FTI_COMM_WORLD, IERROR);
      INVALID = TMP;
   END IF    

   IF ( FTI_APP_RANK == 0 .AND. (STATE == RESTART .OR. STATE == KEEP) ) THEN
      IF ( INVALID == 0 ) THEN
         PRINT *, "[SUCCESSFULL]"
      ELSE
         PRINT *, "[NOT SUCCESSFULL]"
         SCES = 0
      END IF
   END IF

   CALL MPI_BARRIER ( FTI_COMM_WORLD, IERROR );

   CALL FTI_FINALIZE(IERROR)
   CALL MPI_FINALIZE(IERROR)

   IF ( SCES == 1 ) THEN 
      CALL EXIT ( 0 )
   ELSE
      CALL EXIT ( DATA_CORRUPT )
   END IF
END PROGRAM

